### PR TITLE
[qualys_vmdr] add user activity log data stream

### PR DESCRIPTION
**Pull request by andrewkroh**
## Proposed commit message

Add a new data stream to pull user activity logs from Qualys VMDR. It collects data from the `/api/2.0/fo/activity_log/?action=list` API. The API returns data in CSV format.

https://docs.qualys.com/en/vm/api/users/index.htm#t=activity%2Fexport_activity.htm

## Checklist

- [ ] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ] 

## Related issues

-

### CODE REVIEW

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    "GET",
+    state.url + "/api/2.0/fo/activity_log/?" + {
+      "action": ["list"],
+      "since_datetime": [string(state.?cursor.?latest_ts.orValue(now - duration(state.initial_interval)))],
+      "truncation_limit": [string(state.batch_size)],
+    }.format_query()
+  ).with({
+    "Header":{
+      "Authorization": ["Basic " + (state.user + ":" + state.password).base64()],
+      "X-Requested-With": ["Elastic-Agent"],
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:36]: Is this header required? If not, then the built in Basic Auth could be used.

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  }).do_request().as(resp, (
+    resp.StatusCode == 200
+    ?
+      bytes(resp.Body).mime("text/csv; header=present").as(events, state.with({
+          "events": events.map(e, {"message": e.encode_json()}),
+          "want_more": size(events) >= state.batch_size,
+          "cursor": {
+              ?"latest_ts": size(events) > 0 ?
+                optional.of(events.map(e, e.Date.parse_time(time_layout.RFC3339)).max()) :
+                state.?cursor.?latest_ts,
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:47]:
```suggestion
                optional.of(events.map(e, e.Date.parse_time(time_layout.RFC3339)).max())
              :
                state.?cursor.latest_ts,
```
(after the first `.?`, all subsequent `.` are `.?`)

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    "Header":{
+      "Authorization": ["Basic " + (state.user + ":" + state.password).base64()],
+      "X-Requested-With": ["Elastic-Agent"],
+    }
+  }).do_request().as(resp, (
+    resp.StatusCode == 200
+    ?
+      bytes(resp.Body).mime("text/csv; header=present").as(events, state.with({
+          "events": events.map(e, {"message": e.encode_json()}),
+          "want_more": size(events) >= state.batch_size,
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:43]: How does this work? Is this an optimistic check?

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+{{#each tags as |tag|}}
+  - {{tag}}
+{{/each}}
+{{#contains "forwarded" tags}}
+publisher_pipeline.disable_host: true
+{{/contains}}
+{{#if processors}}
+processors:
+{{processors}}
+{{/if}}
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:75]: final newlines

**Review by efd6 — COMMENTED** [review]
Needs `elastic-package build`.

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs (hunk 1 above)
Comment (andrewkroh) [reply packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:36]: I believe it is required. The docs show

> curl -u "username:password" - "X-Requested-With:curl" "https://<qualys_base_url>/api/2.0/fo/activity_log/?action=list"

The other data streams are setting it to `curl`. Maybe I should just use that for consistency?

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs (hunk 3 above)
Comment (andrewkroh) [reply packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:43]: My thinking is if the response contains the `truncation_limit` count, then it must fetch again to find out if there is more data available. If it receives less than the limit then there must not be any more data to fetch.

Comment (efd6) [reply packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:43]: Thanks, that's what I'm calling an "optimistic check" approach.

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  batch_size: {{batch_size}}
+redact:
+  fields:
+    - password
+program: |
+  request(
+    "GET",
+    state.url + "/api/2.0/fo/activity_log/?" + {
+      "action": ["list"],
+      "since_datetime": [string(state.?cursor.?latest_ts.orValue(now - duration(state.initial_interval)))],
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:30]:
```suggestion
      "since_datetime": [string(state.?cursor.latest_ts.orValue(now - duration(state.initial_interval)))],
```

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  password: {{escape_string password}}
+  initial_interval: {{escape_string initial_interval}}
+  batch_size: {{batch_size}}
+redact:
+  fields:
+    - password
+program: |
+  request(
+    "GET",
+    state.url + "/api/2.0/fo/activity_log/?" + {
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:28]:
```suggestion
    state.url.trim_right("/") + "/api/2.0/fo/activity_log/?" + {
```

File: packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -0,0 +1,176 @@
+---
+description: Pipeline for processing Qualys User Activity Log data.
+processors:
+  - set:
+      field: ecs.version
+      value: 8.11.0
+  - fail:
+      if: ctx.error != null
+      tag: fail_on_cel_error
+      message: CEL program returned an error. Skipping ingest pipeline execution.
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml:10]: We should include this in the package review doc.

File: packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      value: api
+      allow_duplicates: false
+  - append:
+      if: ctx.event?.action == "request"
+      field: event.type
+      value: info
+      allow_duplicates: false
+  - append:
+      if: >
+        ['set', 'add', 'create'].contains(ctx.event?.action)
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml:152]: We're avoiding these transient array constructions for GC reasons.

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+{{/if}}
+{{#if http_client_timeout}}
+resource.timeout: {{http_client_timeout}}
+{{/if}}
+resource.url: {{url}}
+state:
+  user: {{escape_string username}}
+  password: {{escape_string password}}
+  initial_interval: {{escape_string initial_interval}}
+  batch_size: {{batch_size}}
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:21]:
```suggestion
{{#if batch_size}}
  batch_size: {{batch_size}}
{{/if}}
```

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+redact:
+  fields:
+    - password
+program: |
+  request(
+    "GET",
+    state.url + "/api/2.0/fo/activity_log/?" + {
+      "action": ["list"],
+      "since_datetime": [string(state.?cursor.?latest_ts.orValue(now - duration(state.initial_interval)))],
+      "truncation_limit": [string(state.batch_size)],
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:31]:
```suggestion
      ?"truncation_limit": has(state.batch_size) ? optional.of([string(state.batch_size)]) : optional.none(),
```

File: packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    "Header":{
+      "Authorization": ["Basic " + (state.user + ":" + state.password).base64()],
+      "X-Requested-With": ["Elastic-Agent"],
+    }
+  }).do_request().as(resp, (
+    resp.StatusCode == 200
+    ?
+      bytes(resp.Body).mime("text/csv; header=present").as(events, state.with({
+          "events": events.map(e, {"message": e.encode_json()}),
+          "want_more": size(events) >= state.batch_size,
```
Comment (efd6) [inline packages/qualys_vmdr/data_stream/user_activity/agent/stream/cel.yml.hbs:43]:
```suggestion
          "want_more": has(state.batch_size) && size(events) >= state.batch_size,
```
(these changes to batch_size are to satisfy the documented behaviour of "0 is no limit")

File: packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (andrewkroh) [reply packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml:10]: Added this concept to https://docs.google.com/document/d/1lyjfjWyKcpXWAMtxirpnkF9p_uEYSQMChxVcQFHFpPA/edit#heading=h.m2ao71h2bz35

File: packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml (hunk 8 above)
Comment (andrewkroh) [reply packages/qualys_vmdr/data_stream/user_activity/elasticsearch/ingest_pipeline/default.yml]: Fixed 905e3c8

**Review by efd6 — APPROVED** [review]

**Pull request merged**